components/TextFields/src/MDCInputTextField.h

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <UIKit/UIKit.h>
+
+/**
+ A set of Contained Input View states outlined in the Material guidelines. These states overlap with
+ and extend UIControlState.
+ */
+typedef NS_OPTIONS(NSInteger, MDCContainedInputViewState) {
+  /**
+   The default state of the contained input view.
+   */
+  MDCContainedInputViewStateNormal = 1 << 0,
+  /**
+   The state the view is in during normal editing.
+   */
+  MDCContainedInputViewStateFocused = 1 << 1,
+  /**
+   This state most closely resembles the @c selected UIControlState.
+   */
+  MDCContainedInputViewStateActivated = 1 << 2,
+  /**
+   The error state.
+   */
+  MDCContainedInputViewStateErrored = 1 << 3,
+  /**
+   The disabled state.
+   */
+  MDCContainedInputViewStateDisabled = 1 << 4,
+};
+
+/**
+ A UITextField subclass that attempts to do the following:
+
+ - Earnestly interpret and actualize the Material guidelines for text fields, which can be found
+ here: https://material.io/design/components/text-fields.html#outlined-text-field
+
+ - Feel intuitive for someone used to the conventions of iOS development and UIKit controls.
+
+ - Enable easy set up and reliable and predictable behavior.
+
+ */
+@interface MDCInputTextField : UITextField
+
+/**
+ The @c floatingLabel is a label that occupies the area the text usually occupies when there is no
+ text and that floats above the text once there is text. It is distinct from the placeholder.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *floatingLabel;
```

Comment:
Nit/naming: Please call this `label` if it cannot always float.

Reply:
Tip: This sounds more like a placeholder label, that can float when "placeholderLabelFloats" is set to true.

Reply:
It's distinct from a placeholder. The floating label can be floating above the text area and a placeholder can be visible. This is either something that MDCTextField got wrong or that changed in the spec since MDCTextField was written. I confirmed it with Phil in design and clients have actually filed bugs about it.

Reply:
Not terribly opposed to having it be `label`, I just thought `floatingLabel` was more clear

Reply:
I found it confusing that a `floatingLabel` might not float. In that case, does it act as a placeholder and is replaced by input text or is it always floating?

Reply:
`floatingLabel`, as in the label that floats, if `canFloatingLabelFloat` is set to `YES`. I'm sorry if that was confusing.

Reply:
https://material.io/design/components/text-fields.html#anatomy defines this label as "label", so I think it will make sense to align with the Material terminology and call this `label`.